### ISSUE TITLE
(Multiple) Segmentation faults

### ISSUE THREAD

**Issue by ner0x652** [upx/upx#145]
## What's the problem (or question)?

UPX SEGFAULTs at unpacking.
## What should have happened?

The unpacking routine should abort.
## Do you have an idea for a solution?

Adapt existent code to treat the problematic state.
## How can we reproduce the issue?

1. Pack a file(for ex: ls tool)
2. Compile the source code by instrumenting it with afl-g++
3. Execute instrumented UPX binary under AFL fuzzer and unpack file from (1)
4. 

## Please tell us details about your environment.

* UPX version used (`upx --version`): 
upx 3.94-git-d31947e1f016+
UCL data compression library 1.03
zlib data compression library 1.2.8
LZMA SDK version 4.43
* Host Operating System and version:
Debian GNU/Linux 9 (stretch)
* Host CPU architecture:
 64-bit
* Target Operating System and version:
Debian GNU/Linux 9 (stretch)
* Target CPU architecture:
 64-bit

**Comment by ner0x652** [comment]
I can provide the test cases where the UPX segfaults, if you're interested.

**Comment by jreiser** [comment]
@ner0x652 Yes, please; specific test cases help a lot.  Please upload ("Attach files by ...") a couple of specific files.  Specifying what is <tt>afl-g++</tt> (how to get it, the version you used, ...) also would help.

**Comment by ner0x652** [comment]
`afl-g++` is a wrapper over g++ that permits source code instrumentation. It's part of AFL(American Fuzzy Lop) Fuzzer developed by Michal Zalewski. [AFL](http://lcamtuf.coredump.cx/afl/) I've used AFL version 2.51.

I've included my seed file in case you want to re-fuzz with the same file I've used.
[upx_crashes.zip](https://github.com/upx/upx/files/1461980/upx_crashes.zip)

**Comment by jreiser** [comment]
@ner0x652 Please state the actual command line, and upload an actual input file, which causes upx to experience a SIGSEGV.

**Comment by ner0x652** [comment]
I have provided the files in the previously uploaded `.zip`.
The command line: `./upx -d filename`

**Comment by jreiser** [comment]
https://github.com/upx/upx/files/1461980/upx_crashes.zip contains only one file that was compressed by upx, namely zgrep2.  Both the released upx-3.94 and the current tip of the <tt>devel</tt> branch can de-compress that file correctly with no SIGSEGV.
```
$ unzip -l upx_crashes.zip
Archive:  upx_crashes.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  11-10-2017 08:27   upx_crashes/
     3414  11-10-2017 08:25   upx_crashes/id:000015,sig:11,src:000128,op:havoc,rep:16
     4980  11-10-2017 08:27   upx_crashes/zgrep2
     3452  11-10-2017 08:26   upx_crashes/id:000019,sig:11,src:000128+000233,op:splice,rep:32
     3445  11-10-2017 08:25   upx_crashes/id:000016,sig:11,src:000128,op:havoc,rep:32
---------                     -------
    15291                     5 files
$ sum upx_crashes/zgrep2
63695     5
$ cp upx_crashes/zgrep2 foo
$ upx-3.94 -d foo

                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2017
UPX 3.94        Markus Oberhumer, Laszlo Molnar & John Reiser   May 12th 2017

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
      5938 <-      4980   83.87%  linux.sh/i386  foo

Unpacked 1 file.
$ sum foo
42173     6
$ rm foo; cp upx_crashes/zgrep2 foo; upx-devel -d foo
$ devel/upx.out -d foo
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2017
UPX git-58771e+ Markus Oberhumer, Laszlo Molnar & John Reiser   May 13th 2017

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
      5938 <-      4980   83.87%  linux.sh/i386  foo

Unpacked 1 file.
WARNING: this is an unstable beta version - use for testing only! Really.
$ sum foo
42173     6
```

**Comment by jreiser** [comment]
[This is why I ask for the <b>literal</b> command line.]  Aha, those <tt>id:...</tt> files are also compressed by upx.  The tip d14a2c5282050dd6294b899a84e5f4ae88c374d2 of <tt>devel</tt> branch already diagnoses those as <tt>CantUnpackException: bad e_phoff</tt>.

**Comment by ner0x652** [comment]
Yes, as I said, zgrep2 was used as seed file. The file names starting with `id` are the actual files that segfault upx.

Ok, good to know that the issues are already solved. You can close this issue.

**Comment by jreiser** [comment]
The fix ef336dbcc6dc8344482f8cf6c909ae96c3286317 (2017-10-02) has been back-ported to upx-3.94-1.fc26.x86_64.rpm, for example. https://github.com/upx/upx/issues/128
